File: android/DateCalcualtor/src/jp/teres/numa/termcalc/SelectDate.scala
package jp.teres.numa.termcalc

import android.app.{Fragment, FragmentManager}
import android.view._
import android.widget._
import android.os.Bundle

import java.util.{Calendar, Locale}

import jp.teres.numa.DateCalculator._
import jp.teres.numa.utils.Conversions._

import ajd4jp.AJD

abstract class SelectDate extends Fragment with DatePicker.OnDateChangedListener{

	var mPicker : Option[DatePicker] = _

	override def onCreateView(inflater : LayoutInflater, container : ViewGroup, savedInstance : Bundle) : View = {
		val content = inflater.inflate(R.layout.select_data, null)
		val title = Option(getArguments()).map(_.getInt("title"))
		val text = content.findViewById(R.id.title).asOpt[TextView]

		for {
			ti <- title
			te <- text
		} yield {
			te.setText(ti)
		}

		mPicker = content.findViewById(R.id.date).asOpt[DatePicker].map {picker => 
			val today = Calendar.getInstance(Locale.JAPAN)
			val y = today.get(Calendar.YEAR)
			val m = today.get(Calendar.MONTH)
			val d = today.get(Calendar.DAY_OF_MONTH)
			picker.init(y, m, d, this)
			picker
		}
		
		content
	}

	override def onDateChanged(picker : DatePicker, year : Int, monthOfYear : Int, dayOfMonth : Int) : Unit = {}

	val ajd = () => {
		mPicker.map {p =>
			new AJD(p.getYear, p.getMonth + 1, p.getDayOfMonth)
		}
	}
}

class StartDate extends SelectDate {

	override def onResume : Unit = {
		super.onResume
		Option(getActivity).collect {case a : MainActivity => a}
						   .foreach {_.startDate = ajd}		
	}

}
object StartDate extends SelectDate{

	def apply() = {
		val f = new StartDate
		val args = new Bundle()
		args.putInt("title", R.string.start_date)
		f.setArguments(args)

		f
	}
}

class EndDate extends SelectDate {

	override def onResume : Unit = {
		super.onResume
		Option(getActivity).collect { case a : MainActivity => a}
	 					   .foreach {_.endDate = ajd}		
	}

}
object EndDate extends SelectDate {

	def apply() = {
		val f = new EndDate
		val args = new Bundle()
		args.putInt("title", R.string.end_date)
		f.setArguments(args)

		f
	}
}

File: android/app/src/main/java/jp/teres/numa/DateCalculator/jp/teres/numa/tercalc/SelectDate.java
package jp.teres.numa.DateCalculator.jp.teres.numa.tercalc;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;
import android.widget.TextView;

import jp.teres.numa.DateCalculator.R;

abstract public class SelectDate extends Fragment implements DatePicker.OnDateChangedListener{
    public enum SelectDateArgs {
        Start(R.string.title_start_date),
        End(R.string.title_end_date),
        ;
        final int titleId;

        SelectDateArgs(int titleId) {
            this.titleId = titleId;
        }
    }
    private static final String ARGS_TITLE = SelectDate.class.getSimpleName() + ".ARGS_TITLE";

    public static SelectDate createInstance(SelectDateArgs args) {
        final Bundle fragmentArgs = new Bundle();
        fragmentArgs.putInt(SelectDate.ARGS_TITLE, args.titleId);

        final SelectDate fragment;
        if (args == SelectDateArgs.Start) {
            fragment = new StartDate();
        } else {
            fragment = new EndDate();
        }

        fragment.setArguments(fragmentArgs);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View content = inflater.inflate(R.layout.select_data, null);

        final String title;
        if (getArguments() != null) {
            final int resId = getArguments().getInt(SelectDate.ARGS_TITLE);
            if (resId > 0) {
                title = getString(resId);
            } else {
                title = "";
            }
        } else {
            title = "";
        }

        final TextView textView = (TextView)content.findViewById(R.id.title);
        textView.setText(title);

        return content;
    }

    @Override
    public void onDateChanged(DatePicker datePicker, int i, int i2, int i3) {}

    public static class StartDate extends SelectDate{

        private StartDate() {}
    }

    public static class EndDate extends SelectDate{

        private EndDate() {}
    }

}
